modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFileEntryMVCActionCommand.java

```
@@ -57,12 +66,43 @@ protected void doProcessAction(
 		ServiceContext serviceContext = ServiceContextFactory.getInstance(
 			DLFileEntry.class.getName(), actionRequest);
 
-		_dlAppService.copyFileEntry(
-			fileEntryId, destinationFolderId, destinationRepositoryId,
-			serviceContext);
+		ThemeDisplay themeDisplay = (ThemeDisplay)actionRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
+
+		JSONObject jsonObject = _jsonFactory.createJSONObject();
+
+		try {
+			_dlAppService.copyFileEntry(
+				fileEntryId, destinationFolderId, destinationRepositoryId,
+				serviceContext);
+
+			jsonObject.put("copied", Boolean.TRUE);
+		}
+		catch (PortalException portalException) {
+			jsonObject.put("copied", Boolean.FALSE);
+
+			String errorMessage = themeDisplay.translate(
+				portalException.getMessage());
+
+			jsonObject.put("errorMessage", errorMessage);
+		}
+
+		try {
+			JSONPortletResponseUtil.writeJSON(
+				actionRequest, actionResponse, jsonObject);
+		}
+		catch (IOException ioException) {
+			_log.error(ioException);
+		}
```

Can you refactor this so that there's a single try/catch block? Also, use JSONUtil instead of the json factory and inline the objects:
```
try {
    ...

    JSONPortletResponseUtil.writeJSON(actionRequest, actionResponse, JSONUtil.put(...) ...);
}
catch (PortalException portalException) {
    JSONPortletResponseUtil.writeJSON(..., ..., JSONUtil.put(...) ...);
}
```
As a general rule, try to keep the "happy path" and the error code explicitly separated, and avoid swallowing exceptions (like the IOException there). If you cannot do anything reasonable with an error, it is better to let it fail and bubble up.

---

I've refactored it! :)